pkg/cache/inmemory.go

```diff
@@ -0,0 +1,302 @@
+		ConstLabels: prometheus.Labels{"name": name},
+	})
+
+	c.current = promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+		Name:        "thanos_cache_inmemory_items",
+		Help:        "Current number of items in the inmemory cache.",
+		ConstLabels: prometheus.Labels{"name": name},
+	})
+
+	c.currentSize = promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+		Name:        "thanos_cache_inmemory_items_size_bytes",
+		Help:        "Current byte size of items in the inmemory cache.",
+		ConstLabels: prometheus.Labels{"name": name},
+	})
+
+	c.totalCurrentSize = promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+		Name: "thanos_cache_inmemory_total_size_bytes",
+		Help: "Current byte size of items (both value and key) in the inmemory cache.",
+	})
+
+	_ = promauto.With(reg).NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "thanos_cache_inmemory_max_size_bytes",
+		Help: "Maximum number of bytes to be held in the inmemory cache.",
+	}, func() float64 {
+		return float64(c.maxSizeBytes)
+	})
+	_ = promauto.With(reg).NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "thanos_cache_inmemory_max_item_size_bytes",
+		Help: "Maximum number of bytes for single entry to be held in the inmemory cache.",
+	}, func() float64 {
+		return float64(c.maxItemSizeBytes)
+	})
+
+	// Initialize LRU cache with a high size limit since we will manage evictions ourselves
+	// based on stored size using `RemoveOldest` method.
+	l, err := lru.NewLRU(maxInt, c.onEvict)
+	if err != nil {
+		return nil, err
+	}
+	c.lru = l
+
+	level.Info(logger).Log(
+		"msg", "created in-memory inmemory cache",
+		"maxItemSizeBytes", c.maxItemSizeBytes,
+		"maxSizeBytes", c.maxSizeBytes,
+		"maxItems", "maxInt",
+	)
+	return c, nil
+}
+
+func (c *InMemoryCache) onEvict(key, val interface{}) {
+	keySize := uint64(len(key.(string)))
+	entrySize := uint64(len(val.(cacheDataWithTTLWrapper).data))
+
+	c.evicted.Inc()
+	c.current.Dec()
+	c.currentSize.Sub(float64(entrySize))
+	c.totalCurrentSize.Sub(float64(keySize + entrySize))
+
+	c.curSize -= entrySize
+}
+
+func (c *InMemoryCache) get(key string) ([]byte, bool) {
+	c.requests.Inc()
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	v, ok := c.lru.Get(key)
+	if !ok {
+		return nil, false
+	}
+	// If the present time is greater than the TTL for the object from cache, the object will be
+	// removed from the cache and a nil will be returned
+	if time.Now().After(v.(cacheDataWithTTLWrapper).expiryTime) {
+		c.hitsExpired.Inc()
+		c.lru.Remove(key)
+		return nil, false
+	}
+	c.hits.Inc()
+	return v.(cacheDataWithTTLWrapper).data, true
+}
+
+func (c *InMemoryCache) set(key string, val []byte, ttl time.Duration) {
+	var size = uint64(len(val))
+	keySize := uint64(len(key))
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	if _, ok := c.lru.Get(key); ok {
+		return
+	}
+
+	if !c.ensureFits(size) {
+		c.overflow.Inc()
+		return
+	}
+
+	// The caller may be passing in a sub-slice of a huge array. Copy the data
+	// to ensure we don't waste huge amounts of space for something small.
```

Comment:
Just a thought: Is this a responsibility of cache, or caller? Problem in doing it in cache is that cache must always do a copy, because it doesn't know if it's required or not. Caller can make that distinction however, which is why it may be preferable to not do it here.

Either way, I think this behaviour should be documented at `InMemoryCache` struct.

Reply:
You are right, I feel this is the responsibility of the caller. But to keep things consistent with https://github.com/thanos-io/thanos/blob/1fff9a7157d1f3b7d5e0ce36f999b10bb3ae7ec4/pkg/store/cache/inmemory.go#L222-L223 I will leave a comment in the `InMemoryCache` struct.